File: py/src/practise150/p4_group_anagrams.py
"""
Given an array of strings, group all anagrams together into sublists. You may return the output in any order.

Eg:
    Input: strs = ["act","pots","tops","cat","stop","hat"]

    Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]

    Input: strs = ["x"]

    Output: [["x"]]
"""

from collections import defaultdict


class Solution:
    def isAnagram(self, s: str, t: str):
        if len(s) != len(t):
            return False

        alphabet_frequency = [0 for x in range(26)]

        for i in range(len(s)):
            alphabet_frequency[ord(s[i]) - ord("a")] += 1
            alphabet_frequency[ord(t[i]) - ord("a")] -= 1
        for x in alphabet_frequency:
            if x != 0:
                return False
        return True

    def group_anagrams(self, strs):
        '''
        time: O(n*m)
        '''
        if len(strs) <= 1:
            return [strs]

        mapofanagrams = defaultdict(list)

        for s in strs:
            frequency = [0] * 26
            for i in range(len(s)):
                frequency[ord(s[i]) - ord("a")] += 1

            mapofanagrams[tuple(frequency)].append(s)

        return list(mapofanagrams.values())

File: go/practise150/p4_group_anagrams.go
package practise150

// """
// Given an array of strings, group all anagrams together into sublists. You may return the output in any order.

// Eg:
//     Input: strs = ["act","pots","tops","cat","stop","hat"]

//     Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]

//     Input: strs = ["x"]

//     Output: [["x"]]
// """

func group_anagrams(sliceofstrings []string) [][]string {
	res1 := make(map[[26]int][]string)

	for _, a_string := range sliceofstrings {
		frequency_array := [26]int{}
		for _, rune := range a_string {
			frequency_array[rune-'a']++
		}
		res1[frequency_array] = append(res1[frequency_array], a_string)
	}

	var final [][]string
	for _, val := range res1 {
		final = append(final, val)
	}

	return final
}
